model/riscv_insts_zicsr.sail

```diff
@@ -202,14 +179,19 @@ function clause execute CSR(csr, rs1, rd, is_imm, op) = {
   else if not(ext_check_CSR(csr, cur_privilege, isWrite))
   then { ext_check_CSR_fail(); RETIRE_FAIL }
   else {
-    let csr_val = readCSR(csr); /* could have side-effects, so technically shouldn't perform for CSRW[I] with rd == 0 */
+    let csr_val = read_CSR(csr); /* could have side-effects, so technically shouldn't perform for CSRW[I] with rd == 0 */
     if isWrite then {
       let new_val : xlenbits = match op {
         CSRRW => rs1_val,
         CSRRS => csr_val | rs1_val,
         CSRRC => csr_val & ~(rs1_val)
       };
-      writeCSR(csr, new_val)
+      let final_val = write_CSR(csr, new_val);
```

Comment:
Since there is only one caller this seems fine but I wonder if it makes sense to have a logging variant.

In the future it might be nice to have logging for all implicit CSR reads and writes, but not for this PR.
I was thinking of maybe something like a scattered enum for all the CSRs so you don't have to hardcode the number and then using something like
`let mtvec = CSR(CSR_MTVEC)` could log the implicit read.

Reply:
Yeah we have had to put stuff like `trackCSRWrite(csr_name_map("mstatus"))` in a few places in our code. A scattered enum might make more sense. It would also make the `write_CSR`/`read_CSR` a lot code easier to read.

Something like

```
scattered enum CSR

enum clause CSR = MSTATUS
enum clause CSR = MTVEC
...

scattered mapping csr_name_map : CSR <-> string

mapping class csr_name_map = MSTATUS <-> "mstatus"
mapping class csr_name_map = MTVEC <-> "mtvec"
...

scattered mapping csr_index : CSR <-> bits(12)

mapping clause csr_index = MSTATUS <-> 0x123
mapping clause csr_index = MTVEC <-> 0x456
...

function execute(CSR... = {
  let index : bits(12) = ...
  if not(csr_index_matches(index)) ...
  let csr = csr_index(index);
  write_CSR(csr, ...)
```

I'm not sure I quite follow what you mean by "a logging variant"?

Reply:
I meant I am not sure if we want every read/write of a CSR to be emitted in the logfile since that could end up being really verbose.

Reply:
And by logging variant I meant a version of write_CSR/read_CSR that logs the access as opposed to doing it in the caller.

Reply:
Ah I see, that makes sense.

Reply:
(But I think we should do it in a future PR.)

Reply:
agreed, no need to do it here.